### PR TITLE
Refactor Variable Accesses out of SDB

### LINKED ISSUE #15944: Varsub is very slow
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Archlinux x86 64
| File format of the file you reverse (mandatory)      | ELF 
| Architecture/bits of the file (mandatory)            | x86/64
| r2 -v full output, **not truncated** (mandatory)     | radare2 4.3.0-git 23793 @ linux-x86-64 git.4.2.1-24-gbe10bbe7b commit: be10bbe7b08ebada8eee41bd3844ce383946e3ec build: 2020-02-09__10:57:34

### Expected behavior
Printing disassembly with varsub enabled doesn't take significantly longer than without it.

### Actual behavior

On large function varsub can make commands like agfj 10x slower.

### Steps to reproduce the behavior 
- Open medium size executable like cmake
- `aaa` less analysis might be sufficient, I don't know
- 'aflt nbbs/sort` find biggest functions
- `e asm.var.sub=true` enable varsub
- `?t agfj function_name`  write down the time
- `e asm.var.sub=false` enable varsub
- `?t agfj function_name` compare with previous time

In my case with  with cmake and function containing ~700 bbs and 200 local vars varsub made it almost 10x slower -  2.24s vs 0.28s.

### Observations about code and potential improvements

* List of sp and bp variables gets created for each asm line
* To find the variables which might be used in instruction varsub iterates over all variables instead of using infromation from disassembler
* Variable repalcement is done as string replacement and guessing how disassembler printed reference to the variable

Potential improvements starting with with best and less likely to be done in near time, ending with simpler solutions that could make it less bad until better fix.

* Disassembler returns asm text with annotations describing which segment of text is instruction name, argument 1, argument 2, register name, expression using a register. Disassembler already has this information but currently there is no mapping between it and the text. This would reduce amount of architecture specific assembly text parsing code. 
* Use the structured instruction arguments information from disassembler so that only relevant variables need to be searched in instruction text
* store information about variables in a structure that allows efficient lookup from instruction description without iterating over all of them
* Basic variable list cache so that it doesn't get created from scratch for each instruction

### PR DESCRIPTION
**Pull request by thestr4ng3r**
Fix #15944

### CODE REVIEW

**Comment by thestr4ng3r** [comment]
```
==32261== Conditional jump or move depends on uninitialised value(s)
==32261==    at 0x598CDCE: var_accesses_list (cmd_anal.c:978)
==32261==    by 0x598D017: list_vars (cmd_anal.c:1016)
==32261==    by 0x598DBFA: var_cmd (cmd_anal.c:1199)
==32261==    by 0x5996F44: cmd_anal_fcn (cmd_anal.c:3380)
==32261==    by 0x59AF9B7: cmd_anal (cmd_anal.c:9958)
==32261==    by 0x5A09FD0: r_cmd_call (cmd_api.c:248)
==32261==    by 0x5A0464B: r_core_cmd_subst_i (cmd.c:3767)
==32261==    by 0x5A006E8: r_core_cmd_subst (cmd.c:2685)
==32261==    by 0x5A00834: r_core_cmd_subst (cmd.c:2717)
==32261==    by 0x5A00834: r_core_cmd_subst (cmd.c:2717)
==32261==    by 0x5A00834: r_core_cmd_subst (cmd.c:2717)
==32261==    by 0x5A073A1: run_cmd_depth (cmd.c:6627)
==32261==
```

File: libr/anal/function.c [hunk 1]
```diff
@@ -176,6 +190,27 @@ R_API bool r_anal_function_relocate(RAnalFunction *fcn, ut64 addr) {
 		return false;
 	}
 	ht_up_delete (fcn->anal->ht_addr_fun, fcn->addr);
+
+	// relocate the var accesses (their addrs are relative to the function addr)
+	st64 delta = (st64)addr - (st64)fcn->addr;
```
Comment (radare) [inline libr/anal/function.c:195]: sounds like this wont work with kernel addresses

File: libr/anal/var.c [hunk 2]
```diff
@@ -262,35 +259,59 @@ R_API RAnalVar *r_anal_function_get_var(RAnalFunction *fcn, char kind, int delta
...
-		} else if (v1->kind == R_ANAL_VAR_KIND_SPV) {
-			regname = r_reg_get_name (a->reg, R_REG_NAME_SP);
+	return r_reg_getv (anal->reg, regname) + var->delta;
+}
+
+R_API st64 r_anal_function_get_var_stackptr_at(RAnalFunction *fcn, int delta, ut64 addr) {
+	st64 offset = (st64)addr - (st64)fcn->addr;
+	RPVector *inst_accesses = ht_up_find (fcn->inst_vars, offset, NULL);
+	if (!inst_accesses) {
+		return INT_MAX;
```
Comment (radare) [inline libr/anal/var.c:278]: INT64_MAX

Comment (radare) [reply libr/anal/var.c:278]: err ST64_MAX

File: libr/anal/var.c [hunk 3]
```diff
@@ -262,35 +259,59 @@ R_API RAnalVar *r_anal_function_get_var(RAnalFunction *fcn, char kind, int delta
...
+	}
+	RAnalVar *var = r_pvector_at (inst_accesses, 0);
+	size_t index;
+	r_vector_lower_bound (&var->accesses, offset, index, ACCESS_CMP);
+	RAnalVarAccess *acc = NULL;
+	if (index < var->accesses.len) {
+		acc = r_vector_index_ptr (&var->accesses, index);
+	}
+	if (!acc || acc->offset != offset) {
+		return INT_MAX;
```
Comment (radare) [inline libr/anal/var.c:300]: same

File: libr/anal/var.c [hunk 4]
```diff
@@ -262,35 +259,59 @@ R_API RAnalVar *r_anal_function_get_var(RAnalFunction *fcn, char kind, int delta
...
-	if (v1) {
-		if (v1->kind == R_ANAL_VAR_KIND_BPV) {
-			regname = r_reg_get_name (a->reg, R_REG_NAME_BP);
-		} else if (v1->kind == R_ANAL_VAR_KIND_SPV) {
-			regname = r_reg_get_name (a->reg, R_REG_NAME_SP);
+	return r_reg_getv (anal->reg, regname) + var->delta;
+}
+
+R_API st64 r_anal_function_get_var_stackptr_at(RAnalFunction *fcn, int delta, ut64 addr) {
+	st64 offset = (st64)addr - (st64)fcn->addr;
```
Comment (radare) [inline libr/anal/var.c:275]: looks wrong. that wont work when the higher bit is set

File: libr/core/disasm.c [hunk 5]
```diff
@@ -975,6 +975,7 @@ static char *get_op_ireg (void *user, ut64 addr) {
 }
 
 static int get_ptr_at(void *user, RAnalFunction *fcn, int delta, ut64 addr) {
+	return (int)r_anal_function_get_var_stackptr_at (fcn, delta, addr);
```
Comment (radare) [inline libr/core/disasm.c:978]: dead code and scary cast

File: libr/anal/var.c (hunk 4 above)
Comment (thestr4ng3r) [reply libr/anal/var.c:272]: It seems to work fine. I added tests for exactly these cases.

File: libr/anal/var.c [hunk 6]
```diff
@@ -262,35 +256,58 @@ R_API RAnalVar *r_anal_function_get_var(RAnalFunction *fcn, char kind, int delta
...
+	} else if (var->kind == R_ANAL_VAR_KIND_SPV) {
+		regname = r_reg_get_name (anal->reg, R_REG_NAME_SP);
 	}
-	RAnalVar *v1 = r_anal_function_get_var_byname (fcn, name);
-	if (v1) {
-		if (v1->kind == R_ANAL_VAR_KIND_BPV) {
-			regname = r_reg_get_name (a->reg, R_REG_NAME_BP);
-		} else if (v1->kind == R_ANAL_VAR_KIND_SPV) {
-			regname = r_reg_get_name (a->reg, R_REG_NAME_SP);
+	return r_reg_getv (anal->reg, regname) + var->delta;
```
Comment (radare) [inline libr/anal/var.c:268]: Regname can be null at this point i think. And this will make it assert

File: libr/anal/var.c [hunk 7]
```diff
@@ -346,38 +362,48 @@ struct VarUsedType {
...
-	RAnalFunction *fcn = NULL;
-	if (sdb_fmt_tobin (var_def, SDB_VARUSED_FMT, &vut) == 4) {
-		fcn = r_anal_get_function_at (anal, vut.fcn_addr);
-		if (fcn) {
-			res = r_anal_function_get_var (fcn, vut.type[0], vut.delta);
-		}
-		sdb_fmt_free (&vut, SDB_VARUSED_FMT);
+R_API R_BORROW RPVector *r_anal_function_get_vars_used_at(RAnalFunction *fcn, ut64 op_addr) {
+	r_return_val_if_fail (fcn, NULL);
+	return ht_up_find (fcn->inst_vars, (st64)op_addr - (st64)fcn->addr, NULL);
```
Comment (radare) [inline libr/anal/var.c:367]: Ill probably put that substraction in a local const variable for readability. And im still unhappy with that signed cast but we can work on thst later in a separate pr if we see issues. Thanks for writing the tests

File: libr/include/r_parse.h [hunk 8]
```diff
@@ -32,7 +32,7 @@ typedef struct r_parse_t {
 	// RAnal *anal; // weak anal ref XXX do not use. use analb.anal
 	RList *parsers;
 	RAnalVarList varlist;
-	int (*get_ptr_at)(void *user, RAnalFunction *fcn, int delta, ut64 addr);
+	st64 (*get_ptr_at)(void *user, RAnalFunction *fcn, st64 delta, ut64 addr);
```
Comment (radare) [inline libr/include/r_parse.h:35]: I assume the delta is stack relative.. and i wont expect a 63bit delta for a stackframe

**Pull request merged**